pkg/utils/bpf-iter-ns/bpf-iter-ns.go

```diff
@@ -0,0 +1,197 @@
+		return ReadOnCurrentPidNs(iter)
+	} else {
+		return ReadOnHostPidNs(iter)
+	}
+}
+
+// ReadOnCurrentPidNs reads the iterator in the current pid namespace.
+func ReadOnCurrentPidNs(iter *link.Iter) ([]byte, error) {
+	file, err := iter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("open BPF iterator: %w", err)
+	}
+	defer file.Close()
+	buf, err := io.ReadAll(file)
+	if err != nil {
+		return nil, fmt.Errorf("read BPF iterator: %w", err)
+	}
+	return buf, err
+}
+
+// ReadOnHostPidNs reads the iterator in the host pid namespace.
+// It does so by pinning the iterator in a temporary directory in the host bpffs,
+// and then creating a systemd service that will read the iterator and write it
+// to a temporary pipe. The pipe is then read and returned.
+func ReadOnHostPidNs(iter *link.Iter) ([]byte, error) {
+	selfPidOnHost, err := os.Readlink(host.HostProcFs + "/self")
+	if err != nil {
+		return nil, fmt.Errorf("readlink /proc/self: %w", err)
+	}
+	if selfPidOnHost == "" {
+		return nil, fmt.Errorf("empty /proc/self symlink")
+	}
+
+	tmpPinDir, err := os.MkdirTemp("/sys/fs/bpf", "ig-iter-")
+	if err != nil {
+		return nil, fmt.Errorf("creating temporary directory in bpffs: %w", err)
+	}
+	pinPath := filepath.Join(tmpPinDir, "iter")
+	err = iter.Pin(pinPath)
+	if err != nil {
+		return nil, fmt.Errorf("pinning iterator: %w", err)
+	}
+	defer os.RemoveAll(tmpPinDir)
+
+	r, w, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("creating pipe: %w", err)
+	}
+	writerPath := fmt.Sprintf("/proc/%s/fd/%d", selfPidOnHost, w.Fd())
+
+	var buf []byte
+	var errReader error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		stdoutReader := bufio.NewReader(r)
+		// ReadAll will block until the write-side of the pipe is closed in both processes
+		// (the systemd service and this process)
+		buf, errReader = io.ReadAll(stdoutReader)
+		r.Close()
+		wg.Done()
+	}()
+
+	conn, err := systemdDbus.NewSystemdConnectionContext(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("connecting to systemd: %w", err)
+	}
+	defer conn.Close()
+
+	randomID := make([]byte, 8)
+	for i := 0; i < 8; i++ {
+		randomID[i] = byte(65 + rand.Intn(25))
+	}
+	unitName := fmt.Sprintf("ig-%s.service", string(randomID))
+
+	statusChan := make(chan string, 1)
+	properties := []systemdDbus.Property{
+		systemdDbus.PropDescription("Inspektor Gadget job on host pidns"),
+		// Type=oneshot ensures that StartTransientUnitContext will only return "done" when the job is done
+		systemdDbus.PropType("oneshot"),
+		systemdDbus.PropExecStart([]string{
+			"/bin/sh",
+			"-c",
+			fmt.Sprintf("cat %s > %s", pinPath, writerPath),
+		}, true),
+	}
+	properties = append(properties, systemdDbus.PropDescription("Inspektor Gadget job on host pidns"))
+
+	_, err = conn.StartTransientUnitContext(context.TODO(),
+		unitName, "fail", properties, statusChan)
+	if err != nil {
+		return nil, fmt.Errorf("starting transient unit: %w", err)
+	}
+	timeout := time.NewTimer(5 * time.Second)
```

Comment:
Is this big enough to cover reading a lot of entries in a busy system?

Reply:
I think it should be fast, but I can change it to 10s.